packages/medusa-interfaces/src/base-service.js

```diff
@@ -158,9 +158,35 @@ class BaseService {
    * @param {string} isolation - the isolation level to be used for the work.
    * @return {any} the result of the transactional work
    */
-  async atomicPhase_(work, isolation) {
+  async atomicPhase_(work, isolationOrErrorHandler, maybeErrorHandler) {
+    let errorHandler = maybeErrorHandler
+    let isolation = isolationOrErrorHandler
+    if (typeof isolationOrErrorHandler === "function") {
+      isolation = null
+      errorHandler = isolationOrErrorHandler
+    }
+
     if (this.transactionManager_) {
-      return work(this.transactionManager_)
+      const doWork = async (m) => {
+        this.manager_ = m
+        this.transactionManager_ = m
+        try {
+          const result = await work(m)
+          return result
+        } catch (error) {
+          if (errorHandler) {
+            const queryRunner = this.transactionManager_.queryRunner
+            if (queryRunner.isTransactionActive) {
+              await queryRunner.rollbackTransaction()
+            }
+
+            await errorHandler(error)
+          }
```

Comment:
Very nice

Should we add it to the second definition of `doWork` as well (lines 192-205) since it's also an issue if we run the service without `withTransaction` but with `atomicPhase_` since then `doWork` is still invoked in a transaction and then `errorhandler` does nothing?